From 11bc9cbd12122f242ee0716f8013b99f1cac03da Mon Sep 17 00:00:00 2001
From: Mika Laitio <lamikr@gmail.com>
Date: Fri, 24 May 2024 16:47:18 -0700
Subject: [PATCH 3/7] gradient_builder.cc dangling reference warning fix
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Causes build break on ubuntu and fedora for following:

onnxruntime/orttraining/orttraining/core/graph/gradient_builder.cc: In member function ‘virtual std::vector<onnxruntime::training::NodeDef> onnxruntime::training::GetTransposeGradient::GetGradientDefsImpl() const’:
onnxruntime/orttraining/orttraining/core/graph/gradient_builder.cc:660:29: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  660 |     const TensorShapeProto& input_shape = I(0).type_proto->tensor_type().shape();

Signed-off-by: Mika Laitio <lamikr@gmail.com>
---
 orttraining/orttraining/core/graph/gradient_builder.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/orttraining/orttraining/core/graph/gradient_builder.cc b/orttraining/orttraining/core/graph/gradient_builder.cc
index 22dcf4eb92..81c14a2282 100755
--- a/orttraining/orttraining/core/graph/gradient_builder.cc
+++ b/orttraining/orttraining/core/graph/gradient_builder.cc
@@ -657,7 +657,7 @@ IMPLEMENT_GRADIENT_BUILDER(GetTransposeGradient) {
   auto attributes = SrcNodeAttributes();
   std::vector<AttributeProto> new_attributes;
   if (attributes.empty()) {
-    const TensorShapeProto& input_shape = I(0).type_proto->tensor_type().shape();
+    const TensorShapeProto input_shape = I(0).type_proto->tensor_type().shape();
     if (input_shape.dim_size() > 0) {  // input_shape is available
       size_t n = static_cast<size_t>(input_shape.dim_size()) - 1;
       bw_perm.resize(n + 1);
-- 
2.46.0

